### DISCUSSION TITLE
스캐빈저 위빙 후 Collector 빌드 시 Method 수의 제한 문의

### DISCUSSION THREAD

**Discussion by si1201-lgu** [naver/scavenger#136] [Q&A]
안녕하세요 스캐빈저를 활용하여 오래된 레거시 프로젝트의 데드코드를 식별중에 있습니다.
관련하여 몇번의 시도 끝에 스캐빈저를 성공적으로 빌드시켰습니다.
이렇게 훌륭한 오픈솔루션을 제공해주셔서 우선 감사인사 드립니다.
본론으로 들어가서, 현재 레거시 프로젝트에 데드코드를 식별하기 위하여
프로젝트 전체 Packege를 Config에 설정하니 Method가 3만개 이상이 되어, Collector가 제대로 동작을 하지 않았습니다.
해서 Packege 레벨을 더 낮춰서 설정한 후, Method수를 줄이니 Collector 빌드에 성공을 하였습니다.
이에 문의가 있습니다.
1. Method의 수에 대한 제약조건이 있을까요?
2. 인프라 환경에 따라 다를 수 있겠지만, 현재 AWS 환경에 구축하였는데, AWS환경에서 특별히 설정 해줘야 할 것들이 존재할까요?
3. 3만여개 이상의 Method를 가진 프로젝트의 Packege 전체를 풀로드하지 못하다보니, 제약이 많이 걸려 더 큰 다른 프로젝트에 적용하기에 어려울 것으로 판단되는데 이를 해소할 수 있는 방법이 있을까요?

 스크린샷을 남기지 못해 우선 양해부탁드리겠습니다.
바쁘시겠지만 이와 비슷한 경험이 있으실지 하여 QnA남깁니다.
확인 한번 부탁드리겠습니다.

감사합니다.

**Comment by sohyun-ku** [comment]
@si1201-lgu 

안녕하세요 :)
추가 자료(에러로그, 스크린샷)가 없어 정확한 원인 파악이 어려울 수 있지만, 
agent의 method 수집 과정 중 이슈가 있는 것으로 이해하고 우선 가능한 답변 드립니다.

> 1. Method의 수에 대한 제약조건이 있을까요?

default `maxMethodsCount`는 10만개이며, 사용자가 직접 커스텀할 수 있는 기능을 제공하고 있습니다.
다만 3만개의 메서드를 가진 프로젝트에서 이슈가 발생되었다면, grpc max message size에서 제약이 발생되었을 가능성도 있어 보입니다.
이 경우 `-Darmeria.max-message-size` 옵션을 통해 직접 사이즈를 지정하실 수 있습니다. 
자세한 부분은 가이드 참고 부탁드립니다.
- `maxMethodsCount` 관련 가이드 : https://github.com/naver/scavenger/blob/develop/doc/installation.md#configuration-2
- `-Darmeria.max-message-size` 관련 가이드 : https://github.com/naver/scavenger/blob/develop/doc/installation.md#installation

> 2. 인프라 환경에 따라 다를 수 있겠지만, 현재 AWS 환경에 구축하였는데, AWS환경에서 특별히 설정 해줘야 할 것들이 존재할까요?

말씀주신 것 처럼 각 인프라 환경마다 차이가 있어 답변드리기 어렵습니다. 다만, 문제 발생 시 문의해 주시면 함께 검토해드리겠습니다. 

> 3. 3만여개 이상의 Method를 가진 프로젝트의 Packege 전체를 풀로드하지 못하다보니, 제약이 많이 걸려 더 큰 다른 프로젝트에 적용하기에 어려울 것으로 판단되는데 이를 해소할 수 있는 방법이 있을까요? 

1번 답변 참고 부탁드리며, 
추가로 agent configuration 의 `annotations`, `excludePackages`, `excludeConstructors`, `excludeGetterSetter` 와 같은 설정을 사용해 수집할 메서드를 정제해서 사용하신다면 해소하는 방법 중 하나로 도움이 될 수 있을 것으로 보입니다.


> https://github.com/naver/scavenger/blob/develop/doc/installation.md#build-agent
>
configuration | default | Description
-- | -- | --
excludePackages |   | Specifies which of the subpackages specified in packages you want to exclude from tracking.ex) foo.bar, baz.qux
annotations |   | Target sub methods of the annotated class to scan.ex) @org.springframework.web.bind.annotation.RestController, @org.springframework.stereotype.Controller
additionalPackages |   | Specifies the packages excluded by annotations that you want to include in the scan.ex) io.scavenger.javaagent, io.scavenger.dashboard
excludeConstructors | false | Specifies whether constructors should be excluded from scaning. For Spring applications, we recommend setting this to true to exclude constructors of beans from collection because they are called automatically.
excludeGetterSetter | false | Specifies whether getters and setters should be excluded from scaning. Methods in the form of componentN() that are created by the data class of your Kotlin are also excluded. These are often automatically generated by libraries like Lombok, so we recommend setting them to true to exclude them from collection.

**Reply by si1201-lgu** [reply]
안녕하세요 위에 가이드 주신대로
-Darmeria.max-message-size를 좀 변경하였으나, 아래와 같이 처음 문의드렸던 오류가 발생하여 공유드립니다.
혹시 확인 가능하시면 확인 한번만 부탁드리겠습니다.
감사합니다.

1. 컬렉터에 500MB로 변경
![image](https://github.com/user-attachments/assets/2b189420-fe0e-4786-a215-dd0ef73edefc)
2. 컬렉터에 1GB로 변경
![image](https://github.com/user-attachments/assets/37618788-95f9-4d7b-944a-0faf7579b261)

3. Agent에 max-message-size 변경
![image](https://github.com/user-attachments/assets/8ec64009-d099-4e1f-9ffa-d8700d4d17ba)

**Reply by sohyun-ku** [reply]
@si1201-lgu 

max size를 변경했음에도 동일한 문제가 발생할 경우, 해당 문제가 아닐 확률도 있어 보입니다.
남겨주신 로그만으로는 오류가 발생한 원인을 특정하기에는 어려움이 있어 디버깅 해보시는 것을 권장드립니다.
로컬에서 agent를 디버깅할 수 있는 방법 전달드립니다.
https://github.com/naver/scavenger/blob/develop/doc/debugging.md

추가로 사용하신 버전과 collector를 작동시킨 방법, agent configuration 도 전달해주시면 도움이 될 수 있을 것으로 보입니다.

**Reply by si1201-lgu** [reply]
답변주셔서 감사드립니다.
오류 발생하는 곳이 agent가 아니라, collector인데, agent 디버깅을 하면 되는걸까요?

agent conf는 설정을 특별히 한건 없습니다.
""에 있는 값은 보안적으로 문제가 될지 몰라 실제 값이 아닌 걸로 공유드립니다.

`serverUrl = "Collector 도메인"
apiKey = "api key"
appName = "app 명"
packages = "패키지 경로"
codeBase = /opt/tomcat/latest/webapps/"프로젝트명"/WEB-INF
methodVisibility = private
excludeConstructors = true
environment = prod
appVersion = 0.0.1b`

**Reply by sohyun-ku** [reply]
@si1201-lgu 

전달주신 `[scavenger] codebase pulish failed`는 agent에서 발생하는 오류입니다.
collector를 어떻게 실행시키셨는지 확인 가능할까요?

**Reply by YoungJinJung** [reply]
collector에 걸린 명령어는

`java -Darmeria.max-message-size=1GB -jar app.jar`

app.jar는 collector jar입니다.

**Reply by si1201-lgu** [reply]
안녕하세요 위 현상과 다르게 Agent가 부모 프로젝트(?)의 Tomcat을 중지시키는 현상이 간혈적으로 발생하여 문의드립니다.
전부터 많지는 않고 1달기준 3번정도 Tomcat이 셧다운 되는 현상이 발생하였는데, 스캐빈저 Agent에서 위빙할때 발생하는 문제로 파악을 하였습니다.
혹시 관련하여서 해결법이나, 원인을 아실까요?

`siginfo: si_signo: 11 (SIGSEGV), si_code: 1 (SEGV_MAPERR), si_addr: 0x000000000000000c
AX=0x00000003cce1abc0 is an oop
java.lang.String

klass: 'java/lang/String'
RBX=0x000000007ae986cf is an unknown value
RCX=0x000000008db06406 is an unknown value
RDX=0x00000000a6280110 is an unknown value
RSP=0x00007ffa625c5010 is pointing into the stack for thread: 0x00007ffb0c063000
RBP=0x0000000000000001 is an unknown value
RSI=0x00000003c63fd7c0 is an oop
com.navercorp.scavenger.javaagent.collecting.InvocationRegistry
klass: 'com/navercorp/scavenger/javaagent/collecting/InvocationRegistry'
RDI=0x00000003c3ba9780 is pointing into object: 0x00000003c3ba9750
[C
klass: {type array char}
length: 20`

**Reply by sohyun-ku** [reply]
@si1201-lgu 

남겨주신 JVM 크래시 로그만으로는 정확한 원인을 특정하기 어려운 상황입니다만,
`SIGSEGV`(Segmentation Fault)는 주로 JVM 외부 코드와의 상호작용, 특히 네이티브 코드(JNI)나 JVM 설정과 관련된 문제에서 발생하는 경우가 있다고 합니다.
혹시 사용 중인 애플리케이션에서 JNI 또는 네이티브 라이브러리를 사용하고 있는지 확인 부탁드립니다.

추가로 문제를 분석하기 위해 아래와 같은 추가 정보 공유 부탁드립니다.

- 사용 중인 Agent의 버전
- 애플리케이션이 실행되는 Java 버전
- 사용 중인 JVM 옵션.
- 문제가 발생한 환경
  - 운영체제(예: Ubuntu 20.04, CentOS 7 등)
- JVM 버전(예: OpenJDK 11.0.19)
- 문제가 재현 가능한지 여부

또한, 문제 원인 파악을 위해 Agent를 임시로 비활성화하여 실행해보는 방법도 제안드립니다.

**Reply by si1201-lgu** [reply]
안녕하세요 답변이 좀 늦었습니다.
요청주신 부분 확인하여 회신드립니다.

- 사용 중인 애플리케이션에서 JNI 또는 네이티브 라이브러리 : 없음
- 사용 중인 Agent의 버전 : 어디서 확인해야 하는지 정보 부탁드립니다.
- 애플리케이션이 실행되는 Java 버전 : Java 8
- 사용 중인 JVM 옵션 :
`-Djava.util.logging.config.file=/opt/tomcat/latest/conf/logging.properties -Djava.util.logging.manager=org.apache.juli.ClassLoaderLogManager -Djava.awt.headless=true -Djava.security.egd=file:/dev/./urandom -Denv=dev -Dspring.profiles.active=dev -DWEBSQUARE_HOME=/opt/tomcat/latest/webapps/smrt_port/WEB-INF/classes/websquare_home -Dscavenger.configuration=/opt/tomcat/latest/conf/scavenger.conf -javaagent:/opt/tomcat/latest/lib/scavenger-agent.jar -javaagent:/opt/whatap-agent/whatap.agent-2.2.32.jar -Djdk.tls.ephemeralDHKeySize=2048 -Djava.protocol.handler.pkgs=org.apache.catalina.webresources -Dscavenger.configuration=/opt/tomcat/latest/conf/scavenger.conf -javaagent:/opt/tomcat/latest/lib/scavenger-agent.jar -Xms512M -Xmx1024M -server -XX:+UseG1GC -XX:+PrintGCDetails -XX:+PrintGCDateStamps -Xloggc:/var/log/smrtmbl/gc.log -XX:+HeapDumpOnOutOfMemoryError -XX:HeapDumpPath=/var/log/smrtmbl/java_pid.hprof -Djava.endorsed.dirs=/opt/tomcat/latest/endorsed -classpath /opt/tomcat/latest/bin/bootstrap.jar:/opt/tomcat/latest/bin/tomcat-juli.jar -Dcatalina.base=/opt/tomcat/latest -Dcatalina.home=/opt/tomcat/latest -Djava.io.tmpdir=/opt/tomcat/latest/temp org.apache.catalina.startup.Bootstrap start`

- 문제가 발생한 환경 : Linux
- 운영체제(예: Ubuntu 20.04, CentOS 7 등)
- JVM 버전(예: OpenJDK 11.0.19) : OpenJDK 1.8
- 문제가 재현 가능한지 여부 : 재현 불가

현재는 Agent를 중지한 상태인데, 이상은 없습니다.
겨울이 다가와서 날씨가 많이 추운데 건강 유의하시고
바쁘실테지만 확인 한번 부탁드리겠습니다.
감사합니다.

**Reply by sohyun-ku** [reply]
@si1201-lgu 

전달주신 JVM 옵션에 `-javaagent:/opt/tomcat/latest/lib/scavenger-agent.jar`와 `-Dscavenger.configuration`이 중복으로 설정되어 있습니다.  중복된 항목 제거 후 테스트 부탁드립니다.
Scavenger Agent의 버전은 애플리케이션 부팅 시 출력되는 Scavenger 배너에서 확인하실 수 있습니다.
배너에는 Agent 버전과 함께 기본 설정 정보도 함께 표시되니, 관련 로그를 확인해 주시면 감사하겠습니다.
